sdk/appconfiguration/app-configuration/src/featureFlag.ts

```diff
@@ -0,0 +1,274 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  isJsonFeatureFlagPercentageClientFilter,
+  isJsonFeatureFlagTargetingClientFilter,
+  isJsonFeatureFlagTimeWindowClientFilter,
+  JsonFeatureFlag
+} from "./internal/jsonModels";
+import { ConfigurationSetting, ConfigurationSettingParam } from "./models";
+
+/**
+ * The prefix for feature flags.
+ */
+export const featureFlagPrefix = ".appconfig.featureflag/";
+
+/**
+ * The content type for a FeatureFlag
+ */
+export const featureFlagContentType = "application/vnd.microsoft.appconfig.ff+json;charset=utf-8";
+
+/**
+ * Necessary fields for updating or creating a new feature flag.
+ */
+export interface FeatureFlagParam extends ConfigurationSettingParam {
+  // TODO: can we hoist 'clientFilters' higher and avoid this sub-field? Need to talk to team.
+  /**
+   * A Feature filter consistently evaluates the state of a feature flag.
+   * Our feature management library supports three types of built-in filters: Targeting, TimeWindow, and Percentage.
+   * Custom filters can also be created based on different factors, such as device used, browser types, geographic location, etc.
+   *
+   * [More Info](https://docs.microsoft.com/en-us/azure/azure-app-configuration/howto-feature-filters-aspnet-core)
+   */
+  conditions: {
+    clientFilters: (
+      | FeatureFlagTargetingClientFilter
+      | FeatureFlagTimeWindowClientFilter
+      | FeatureFlagPercentageClientFilter
+      | object
+    )[];
+  };
+  /**
+   * Description of the feature.
+   */
+  description?: string;
+  /**
+   * Display name for the feature to use for display rather than the ID.
+   */
+  displayName: string;
+  /**
+   * Boolean flag to say if the feature flag is enabled.
+   */
+  enabled: boolean;
+}
+
+/**
+ * FeatureFlag represents a configuration setting that stores a feature flag value.
+ *
+ * [More Info](https://docs.microsoft.com/azure/azure-app-configuration/concept-feature-management)
+ */
+export interface FeatureFlag extends FeatureFlagParam, ConfigurationSetting {}
+
+/**
+ * Targeting Client filter for the feature flag configuration setting.
+ *
+ * [More Info](https://docs.microsoft.com/en-us/azure/azure-app-configuration/howto-feature-filters-aspnet-core)
+ */
+export interface FeatureFlagTargetingClientFilter {
+  /**
+   * The name of the feature filter.
+   */
+  name: "Microsoft.Targeting";
+  /**
+   * Parameters that can be passed in the filter.
+   */
+  parameters: {
+    // TODO: can we hoist these values up directly into the filter, rather than having this 'parameters' intermediary.
```

Comment:
> // TODO: can we hoist these values up directly into the filter, rather than having this 'parameters' intermediary.

Python has it this way.. I guess I'll keep it the same in TS for consistency.

@seankane-msft @richardpark-msft @AlexGhiondea

Reply:
Python has it this way because of the lack of service side validation. The returned dictionary is not guaranteed to have a "parameters" key, if it did we could skip the middleman here.

Reply:
Since this is all pretty new I don't think we have to strictly adhere to Python.

Here's the .net version of the feature flags:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.featuremanagement.featurefilters?view=azure-dotnet-preview

They seem to do more hiding of the fields than we would, like in [`PercentageFilter`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.featuremanagement.featurefilters.percentagefilter?view=azure-dotnet-preview) since their models also evaluate the feature flag _but_ their settings classes look a bit more similar to what we're exporting here: [`PercentageFilterSettings`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.featuremanagement.featurefilters.percentagefiltersettings?view=azure-dotnet-preview)

So perhaps we could model it a bit more like that.

Reply:
> Python has it this way because of the lack of service side validation. The returned dictionary is not guaranteed to have a "parameters" key, if it did we could skip the middleman here.

Can you not do it in the SDK itself? 

Meaning.. the user gives you 
`{ name; audience: { users; groups ... }; defaultRolloutPercentage }` 
and you'll modify it in the SDK as 
`{ name; parameters: { audience: { users; groups }; defaultRolloutPercentage } }`

Reply:
@richardpark-msft So, you basically like moving them up, right? (I do like that.)

Reply:
Yes, I think so. The only concerns would be if the AppConfig team would do something that would make us regret it ( for instance, if there was a parameter and then they also had a field hoisted on the object itself that conflicted with it). But given the structure of their own settings objects, I doubt it.

Reply:
(definitely still worth checking with them, however) :)

Reply:
I would keep the schema 1:1 with what customers see in the portal. I don't think the improvement to the experience is measurable to justify making the model different from the JSON customers see in the portal.

Reply:
Thanks @pakrym, that's a great suggestion.
1:1 with the portal would also clear some of my other questions. 👏

Reply:
```suggestion
```